# Twitter Data Cleaning

___

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Setup" data-toc-modified-id="Setup-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Setup</a></span><ul class="toc-item"><li><span><a href="#Import-Libraries" data-toc-modified-id="Import-Libraries-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Import Libraries</a></span></li></ul></li><li><span><a href="#Import-Data" data-toc-modified-id="Import-Data-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Import Data</a></span><ul class="toc-item"><li><span><a href="#Load-searched-tweets" data-toc-modified-id="Load-searched-tweets-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Load searched tweets</a></span></li><li><span><a href="#Load-tweet-deets" data-toc-modified-id="Load-tweet-deets-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Load tweet deets</a></span></li><li><span><a href="#Load-users" data-toc-modified-id="Load-users-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Load users</a></span></li></ul></li><li><span><a href="#Convert-Tweets-To-DataFrames" data-toc-modified-id="Convert-Tweets-To-DataFrames-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Convert Tweets To DataFrames</a></span><ul class="toc-item"><li><span><a href="#Tweets-JSONs-To-DataFrames" data-toc-modified-id="Tweets-JSONs-To-DataFrames-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Tweets JSONs To DataFrames</a></span></li><li><span><a href="#Deets-JSONs-To-DataFrames" data-toc-modified-id="Deets-JSONs-To-DataFrames-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Deets JSONs To DataFrames</a></span></li><li><span><a href="#Users-JSONs-To-DataFrames" data-toc-modified-id="Users-JSONs-To-DataFrames-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Users JSONs To DataFrames</a></span></li></ul></li><li><span><a href="#Merge-DataFrames-Into-One" data-toc-modified-id="Merge-DataFrames-Into-One-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Merge DataFrames Into One</a></span><ul class="toc-item"><li><span><a href="#Convert-ID-Strings-To-Integers" data-toc-modified-id="Convert-ID-Strings-To-Integers-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Convert ID Strings To Integers</a></span></li><li><span><a href="#Merge-&quot;Tweets&quot;-DataFrames-With-&quot;Deets&quot;-DataFrames" data-toc-modified-id="Merge-&quot;Tweets&quot;-DataFrames-With-&quot;Deets&quot;-DataFrames-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Merge "Tweets" DataFrames With "Deets" DataFrames</a></span></li><li><span><a href="#Merge-&quot;Tweets-+-Deets&quot;-DataFrames-With-&quot;Users&quot;-DataFrames" data-toc-modified-id="Merge-&quot;Tweets-+-Deets&quot;-DataFrames-With-&quot;Users&quot;-DataFrames-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Merge "Tweets + Deets" DataFrames With "Users" DataFrames</a></span></li><li><span><a href="#Concat-All-&quot;Merged&quot;-DataFrames-Together-Into-One-Big-DataFrame" data-toc-modified-id="Concat-All-&quot;Merged&quot;-DataFrames-Together-Into-One-Big-DataFrame-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Concat All "Merged" DataFrames Together Into One Big DataFrame</a></span></li></ul></li><li><span><a href="#Column-Cleaning" data-toc-modified-id="Column-Cleaning-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Column Cleaning</a></span><ul class="toc-item"><li><span><a href="#Drop-columns" data-toc-modified-id="Drop-columns-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Drop columns</a></span></li></ul></li><li><span><a href="#Filter-DataFrame-Rows-By-&quot;Startup&quot;-Criteria" data-toc-modified-id="Filter-DataFrame-Rows-By-&quot;Startup&quot;-Criteria-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Filter DataFrame Rows By "Startup" Criteria</a></span></li><li><span><a href="#Imputing-Values-As-Necessary" data-toc-modified-id="Imputing-Values-As-Necessary-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Imputing Values As Necessary</a></span><ul class="toc-item"><li><span><a href="#Figure-out-where-NaN-values-are-in-our-data" data-toc-modified-id="Figure-out-where-NaN-values-are-in-our-data-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Figure out where NaN values are in our data</a></span></li></ul></li><li><span><a href="#Export-Final-DataFrame-To-CSV" data-toc-modified-id="Export-Final-DataFrame-To-CSV-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Export Final DataFrame To CSV</a></span></li></ul></div>

___

## Setup

### Import Libraries

In [7]:
import sys
sys.path.append("..")

In [8]:
# Import necessary python packages and functions
from Py_Files.imports import *
from Py_Files.max_range import *
%matplotlib inline

___

## Import Data

### Load searched tweets

In [1]:
with open('../Data/investing.json') as a:
    tweets_investing = json.load(a)
with open('../Data/mobile_payment.json') as b:
    tweets_mobile_payment = json.load(b)
with open('../Data/mobile_wallet.json') as c:
    tweets_mobile_wallet = json.load(c)
with open('../Data/mobile_banking.json') as d:
    tweets_mobile_banking = json.load(d)
with open('../Data/trading.json') as e:
    tweets_trading = json.load(e)
with open('../Data/fintech.json') as f:
    tweets_fintech = json.load(f)
with open('../Data/financial.json') as g:
    tweets_financial = json.load(g)
with open('../Data/money_management.json') as h:
    tweets_money_management = json.load(h)
with open('../Data/stocks.json') as i:
    tweets_stocks = json.load(i)
with open('../Data/transaction.json') as j:
    tweets_transaction = json.load(j)

### Load tweet deets

In [2]:
with open('../Data/deets_investing.json') as aa:
    deets_investing = json.load(aa)
with open('../Data/deets_mobile_payment.json') as bb:
    deets_mobile_payment = json.load(bb)
with open('../Data/deets_mobile_wallet.json') as cc:
    deets_mobile_wallet = json.load(cc)
with open('../Data/deets_mobile_banking.json') as dd:
    deets_mobile_banking = json.load(dd)
with open('../Data/deets_trading.json') as ee:
    deets_trading = json.load(ee)
with open('../Data/deets_fintech.json') as ff:
    deets_fintech = json.load(ff)
with open('../Data/deets_financial.json') as gg:
    deets_financial = json.load(gg)
with open('../Data/deets_money_management.json') as hh:
    deets_money_management = json.load(hh)
with open('../Data/deets_stocks.json') as ii:
    deets_stocks = json.load(ii)
with open('../Data/deets_transaction.json') as jj:
    deets_transaction = json.load(jj)

### Load users

In [3]:
with open('../Data/user_investing.json') as aaa:
    users_investing = json.load(aaa)
with open('../Data/user_mobile_payment.json') as bbb:
    users_mobile_payment = json.load(bbb)
with open('../Data/user_mobile_wallet.json') as ccc:
    users_mobile_wallet = json.load(ccc)
with open('../Data/user_mobile_banking.json') as ddd:
    users_mobile_banking = json.load(ddd)
with open('../Data/user_trading.json') as eee:
    users_trading = json.load(eee)
with open('../Data/user_fintech.json') as fff:
    users_fintech = json.load(fff)
with open('../Data/user_financial.json') as ggg:
    users_financial = json.load(ggg)
with open('../Data/user_money_management.json') as hhh:
    users_money_management = json.load(hhh)
with open('../Data/user_stocks.json') as iii:
    users_stocks = json.load(iii)
with open('../Data/user_transaction.json') as jjj:
    users_transaction = json.load(jjj)

___

## Convert Tweets To DataFrames

### Tweets JSONs To DataFrames

In [118]:
def make_df_column_1(data_source_json, json_key, df_name, df_column_name):
    new_list = []
    for x in range(len(data_source_json)):
        for y in range(len(data_source_json[x]['statuses'])):
            new_list.append(data_source_json[x]['statuses'][y][json_key])
    df_name[df_column_name] = new_list

In [119]:
def make_df_column_2(data_source_json, json_key_1, json_key_2, df_name, df_column_name):
    new_list = []
    for x in range(len(data_source_json)):
        for y in range(len(data_source_json[x]['statuses'])):
            new_list.append(data_source_json[x]['statuses'][y][json_key_1][json_key_2])
    df_name[df_column_name] = new_list

In [164]:
def make_df_column_3(data_source_json, json_key_1, json_key_2, json_key_3, df_name, df_column_name):
    new_list = []
    for x in range(len(data_source_json)):
        for y in range(len(data_source_json[x]['statuses'])):
            new_list.append(data_source_json[x]['statuses'][y][json_key_1][json_key_2][0][json_key_3])
    df_name[df_column_name] = new_list

In [145]:
def make_df_column_4(data_source_json, json_key_1, json_key_2, json_key_3, json_key_4, df_name, df_column_name):
    new_list = []
    for x in range(len(data_source_json)):
        for y in range(len(data_source_json[x]['statuses'])):
            new_list.append(data_source_json[x]['statuses'][y][json_key_1][json_key_2][0][json_key_3][json_key_4])
    df_name[df_column_name] = new_list

In [181]:
def make_dataframes(df_name, data_source_json):
    
    #use first function for first-order JSON keys (make_df_column_1)
    make_df_column_1(data_source_json, 'created_at', df_name, 'tweets_created_at')
    make_df_column_1(data_source_json, 'id', df_name, 'tweets_id')
    make_df_column_1(data_source_json, 'source', df_name, 'tweets_source')
    make_df_column_1(data_source_json, 'in_reply_to_status_id', df_name, 'tweets_in_reply_to_status_id')
    make_df_column_1(data_source_json, 'in_reply_to_user_id', df_name, 'tweets_in_reply_to_user_id')
    make_df_column_1(data_source_json, 'in_reply_to_screen_name', df_name, 'tweets_in_reply_to_screen_name')
    make_df_column_1(data_source_json, 'coordinates', df_name, 'tweets_coordinates')
    make_df_column_1(data_source_json, 'geo', df_name, 'tweets_geo')
    make_df_column_1(data_source_json, 'place', df_name, 'tweets_place')
    make_df_column_1(data_source_json, 'contributors', df_name, 'tweets_contributors')
    make_df_column_1(data_source_json, 'is_quote_status', df_name, 'tweets_is_quote_status')
    make_df_column_1(data_source_json, 'favorited', df_name, 'tweets_favorited')
    make_df_column_1(data_source_json, 'favorite_count', df_name, 'tweets_favorite_count')
    make_df_column_1(data_source_json, 'retweeted', df_name, 'tweets_retweeted')
    make_df_column_1(data_source_json, 'retweet_count', df_name, 'tweets_retweet_count')

    #use second function for second-order JSON keys (make_df_column_2)
    make_df_column_2(data_source_json, 'user', 'id', df_name, 'tweets_user_id')
    make_df_column_2(data_source_json, 'user', 'name', df_name, 'tweets_user_name')
    make_df_column_2(data_source_json, 'user', 'screen_name', df_name, 'tweet_user_screen_name')
    make_df_column_2(data_source_json, 'user', 'location', df_name, 'tweets_user_location')
    make_df_column_2(data_source_json, 'user', 'description', df_name, 'tweets_user_description')
    make_df_column_2(data_source_json, 'user', 'friends_count', df_name, 'tweets_user_friends_count')
    make_df_column_2(data_source_json, 'user', 'followers_count', df_name, 'tweets_user_followers_count')
    make_df_column_2(data_source_json, 'user', 'favourites_count', df_name, 'tweets_user_favourites_count')
    make_df_column_2(data_source_json, 'user', 'geo_enabled', df_name, 'tweets_user_geo_enabled')
    make_df_column_2(data_source_json, 'user', 'verified', df_name, 'tweets_user_verified')
    make_df_column_2(data_source_json, 'user', 'statuses_count', df_name, 'tweets_user_statuses_count')
    make_df_column_2(data_source_json, 'user', 'profile_background_color', df_name, 'tweets_user_profile_background_color')
    make_df_column_2(data_source_json, 'user', 'profile_text_color', df_name, 'tweets_user_profile_text_color')
    make_df_column_2(data_source_json, 'user', 'profile_use_background_image', df_name, 'tweets_user_profile_use_background_image')
    make_df_column_2(data_source_json, 'user', 'has_extended_profile', df_name, 'tweets_user_has_extended_profile')
    make_df_column_2(data_source_json, 'user', 'default_profile', df_name, 'tweets_user_default_profile_image')
    make_df_column_2(data_source_json, 'user', 'following', df_name, 'tweets_user_following')
    make_df_column_2(data_source_json, 'user', 'follow_request_sent', df_name, 'tweets_user_follow_request_sent')
    make_df_column_2(data_source_json, 'user', 'notifications', df_name, 'tweets_user_notifications')

    make_df_column_2(data_source_json, 'entities', 'symbols', df_name, 'tweets_entities_symbols')
    make_df_column_2(data_source_json, 'entities', 'user_mentions', df_name, 'tweets_entities_user_mentions')
    make_df_column_2(data_source_json, 'entities', 'urls', df_name, 'tweets_entities_urls')

In [182]:
tweets_investing_df = pd.DataFrame()
tweets_mobile_payment_df = pd.DataFrame()
tweets_mobile_wallet_df = pd.DataFrame()
tweets_mobile_banking_df = pd.DataFrame()
tweets_trading_df = pd.DataFrame()
tweets_fintech_df = pd.DataFrame()
tweets_financial_df = pd.DataFrame()
tweets_money_management_df = pd.DataFrame()
tweets_stocks_df = pd.DataFrame()
tweets_transaction_df = pd.DataFrame()

In [183]:
make_dataframes(tweets_investing_df, tweets_investing)
make_dataframes(tweets_mobile_payment_df, tweets_mobile_payment)
make_dataframes(tweets_mobile_wallet_df, tweets_mobile_wallet)
make_dataframes(tweets_mobile_banking_df, tweets_mobile_banking)
make_dataframes(tweets_trading_df, tweets_trading)
make_dataframes(tweets_fintech_df, tweets_fintech)
make_dataframes(tweets_financial_df, tweets_financial)
make_dataframes(tweets_money_management_df, tweets_money_management)
make_dataframes(tweets_stocks_df, tweets_stocks)
make_dataframes(tweets_transaction_df, tweets_transaction)

In [184]:
#test a dataframe to make sure it worked:
tweets_investing_df.head(1)

tweets_created_at            tweets_id                                      tweets_source  tweets_in_reply_to_status_id  tweets_in_reply_to_user_id tweets_in_reply_to_screen_name tweets_coordinates tweets_geo tweets_place tweets_contributors  tweets_is_quote_status  tweets_favorited  tweets_favorite_count  tweets_retweeted  tweets_retweet_count  tweets_user_id tweets_user_name tweet_user_screen_name tweets_user_location tweets_user_description  tweets_user_friends_count  tweets_user_followers_count  tweets_user_favourites_count  tweets_user_geo_enabled  tweets_user_verified  tweets_user_statuses_count tweets_user_profile_background_color tweets_user_profile_text_color  tweets_user_profile_use_background_image  tweets_user_has_extended_profile  tweets_user_default_profile_image tweets_user_following tweets_user_follow_request_sent tweets_user_notifications tweets_entities_symbols tweets_entities_user_mentions tweets_entities_urls
0  Fri Jan 31 21:03:41 +0000 2020  1223351165523578885  <a href="http://twitter.com/download/iphone" r...                           NaN                         NaN                           None               None       None         None                None                   False             False                  12761             False                  3014       284278132    Morgan Housel           morganhousel                                  @collabfund                        487                       129374                         54543                    False                  True                       14545                               131516                         333333                                      True                              True                              False                  None                            None                      None                      []                            []                   []

### Deets JSONs To DataFrames

In [194]:
def make_df_column_1_deets(data_source_json, json_key, df_name, df_column_name):
    new_list = []
    for x in range(len(data_source_json)):
        for y in range(len(data_source_json[x]['data'])):
            new_list.append(data_source_json[x]['data'][y][json_key])
    df_name[df_column_name] = new_list

In [195]:
def make_df_column_2_deets(data_source_json, json_key_1, json_key_2, df_name, df_column_name):
    new_list = []
    for x in range(len(data_source_json)):
        for y in range(len(data_source_json[x]['data'])):
            new_list.append(data_source_json[x]['data'][y][json_key_1][json_key_2])
    df_name[df_column_name] = new_list

In [204]:
def make_dataframes_deets(df_name, data_source_json):
    
    #use first function for first-order JSON keys (make_df_column_1)
    make_df_column_1_deets(data_source_json, 'author_id', df_name, 'deets_author_id')
    make_df_column_1_deets(data_source_json, 'created_at', df_name, 'deets_created_at')
    make_df_column_1_deets(data_source_json, 'id', df_name, 'deets_tweet_id')
    make_df_column_1_deets(data_source_json, 'lang', df_name, 'deets_lang')
    make_df_column_1_deets(data_source_json, 'possibly_sensitive', df_name, 'deets_possibly_sensitive')
    make_df_column_1_deets(data_source_json, 'source', df_name, 'deets_source')
    make_df_column_1_deets(data_source_json, 'text', df_name, 'deets_text')

    #use second function for second-order JSON keys (make_df_column_2)
    make_df_column_2_deets(data_source_json, 'stats', 'reply_count', df_name, 'deets_reply_count')
    make_df_column_2_deets(data_source_json, 'stats', 'like_count', df_name, 'deets_like_count')
    make_df_column_2_deets(data_source_json, 'stats', 'quote_count', df_name, 'deets_quote_count')

In [205]:
deets_investing_df = pd.DataFrame()
deets_mobile_payment_df = pd.DataFrame()
deets_mobile_wallet_df = pd.DataFrame()
deets_mobile_banking_df = pd.DataFrame()
deets_trading_df = pd.DataFrame()
deets_fintech_df = pd.DataFrame()
deets_financial_df = pd.DataFrame()
deets_money_management_df = pd.DataFrame()
deets_stocks_df = pd.DataFrame()
deets_transaction_df = pd.DataFrame()

In [206]:
make_dataframes_deets(deets_investing_df, deets_investing)
make_dataframes_deets(deets_mobile_payment_df, deets_mobile_payment)
make_dataframes_deets(deets_mobile_wallet_df, deets_mobile_wallet)
make_dataframes_deets(deets_mobile_banking_df, deets_mobile_banking)
make_dataframes_deets(deets_trading_df, deets_trading)
make_dataframes_deets(deets_fintech_df, deets_fintech)
make_dataframes_deets(deets_financial_df, deets_financial)
make_dataframes_deets(deets_money_management_df, deets_money_management)
make_dataframes_deets(deets_stocks_df, deets_stocks)
make_dataframes_deets(deets_transaction_df, deets_transaction)

In [211]:
deets_investing_df.head(1)

deets_author_id          deets_created_at       deets_tweet_id deets_lang  deets_possibly_sensitive                                       deets_source                           deets_text  deets_reply_count  deets_like_count  deets_quote_count
0       284278132  2020-01-31T21:03:41.000Z  1223351165523578885         en                     False  <a href="http://twitter.com/download/iphone" r...  Investing.  https://t.co/qQQtVyOLid                195             14083                444

### Users JSONs To DataFrames

In [227]:
def make_df_column_1_users(data_source_json, json_key, df_name, df_column_name):
    new_list = []
    for x in range(len(data_source_json)):
        for y in range(len(data_source_json[x]['data'])):
            new_list.append(data_source_json[x]['data'][y][json_key])
    df_name[df_column_name] = new_list

In [228]:
def make_df_column_2_users(data_source_json, json_key_1, json_key_2, df_name, df_column_name):
    new_list = []
    for x in range(len(data_source_json)):
        for y in range(len(data_source_json[x]['data'])):
            new_list.append(data_source_json[x]['data'][y][json_key_1][json_key_2])
    df_name[df_column_name] = new_list

In [252]:
def make_df_column_4_users(data_source_json, json_key_1, json_key_2, json_key_3, json_key_4, df_name, df_column_name):
    new_list = []
    for x in range(len(data_source_json)):
        for y in range(len(data_source_json[x]['data'])):
            try:
                new_list.append(data_source_json[x]['data'][y][json_key_1][json_key_2][json_key_3][0][json_key_4])
            except: new_list.append('')
            continue
    df_name[df_column_name] = new_list

In [253]:
def make_dataframes_users(df_name, data_source_json):
    
    #use first function for first-order JSON keys (make_df_column_1)
    make_df_column_1_users(data_source_json, 'id', df_name, 'users_id')
    make_df_column_1_users(data_source_json, 'created_at', df_name, 'users_created_at')
    make_df_column_1_users(data_source_json, 'description', df_name, 'users_description')
    make_df_column_1_users(data_source_json, 'name', df_name, 'users_name')
    make_df_column_1_users(data_source_json, 'username', df_name, 'users_username')
    make_df_column_1_users(data_source_json, 'verified', df_name, 'users_verified')
    make_df_column_1_users(data_source_json, 'protected', df_name, 'users_protected')
    make_df_column_1_users(data_source_json, 'url', df_name, 'users_url')

    #use second function for second-order JSON keys (make_df_column_2)
    make_df_column_2_users(data_source_json, 'stats', 'followers_count', df_name, 'users_followers_count')
    make_df_column_2_users(data_source_json, 'stats', 'following_count', df_name, 'users_following_count')
    make_df_column_2_users(data_source_json, 'stats', 'tweet_count', df_name, 'users_tweet_count')
    make_df_column_2_users(data_source_json, 'stats', 'listed_count', df_name, 'users_listed_count')
    
    #use third function for fourth-order JSON keys (make_df_column_3)
    make_df_column_4_users(data_source_json, 'entities', 'url', 'urls', 'start', df_name, 'users_url_start')
    make_df_column_4_users(data_source_json, 'entities', 'url', 'urls', 'end', df_name, 'users_url_end')
    make_df_column_4_users(data_source_json, 'entities', 'url', 'urls', 'expanded_url', df_name, 'users_expanded_url')
    make_df_column_4_users(data_source_json, 'entities', 'url', 'urls', 'display_url', df_name, 'users_display_url')

In [254]:
users_investing_df = pd.DataFrame()
users_mobile_payment_df = pd.DataFrame()
users_mobile_wallet_df = pd.DataFrame()
users_mobile_banking_df = pd.DataFrame()
users_trading_df = pd.DataFrame()
users_fintech_df = pd.DataFrame()
users_financial_df = pd.DataFrame()
users_money_management_df = pd.DataFrame()
users_stocks_df = pd.DataFrame()
users_transaction_df = pd.DataFrame()

In [255]:
make_dataframes_users(users_investing_df, users_investing)
make_dataframes_users(users_mobile_payment_df, users_mobile_payment)
make_dataframes_users(users_mobile_wallet_df, users_mobile_wallet)
make_dataframes_users(users_mobile_banking_df, users_mobile_banking)
make_dataframes_users(users_trading_df, users_trading)
make_dataframes_users(users_fintech_df, users_fintech)
make_dataframes_users(users_financial_df, users_financial)
make_dataframes_users(users_money_management_df, users_money_management)
make_dataframes_users(users_stocks_df, users_stocks)
make_dataframes_users(users_transaction_df, users_transaction)

___

## Merge DataFrames Into One

- Merge users with tweets.
- Merge deets with users + tweets
- Done!

### Convert ID Strings To Integers

- **Convert strings to int**
- deets_tweet_id and users_id columns are strings...need to convert to int before merging with "tweets" DataFrames.

In [275]:
#deets_tweet_id is a string...convert to int before merge

deets_investing_df['deets_tweet_id'] = deets_investing_df['deets_tweet_id'].astype(int)
deets_mobile_payment_df['deets_tweet_id'] = deets_mobile_payment_df['deets_tweet_id'].astype(int)
deets_mobile_wallet_df['deets_tweet_id'] = deets_mobile_wallet_df['deets_tweet_id'].astype(int)
deets_mobile_banking_df['deets_tweet_id'] = deets_mobile_banking_df['deets_tweet_id'].astype(int)
deets_trading_df['deets_tweet_id'] = deets_trading_df['deets_tweet_id'].astype(int)
deets_fintech_df['deets_tweet_id'] = deets_fintech_df['deets_tweet_id'].astype(int)
deets_financial_df['deets_tweet_id'] = deets_financial_df['deets_tweet_id'].astype(int)
deets_money_management_df['deets_tweet_id'] = deets_money_management_df['deets_tweet_id'].astype(int)
deets_stocks_df['deets_tweet_id'] = deets_stocks_df['deets_tweet_id'].astype(int)
deets_transaction_df['deets_tweet_id'] = deets_transaction_df['deets_tweet_id'].astype(int)

In [281]:
#users_id is a string...convert to int before merge

users_investing_df['users_id'] = users_investing_df['users_id'].astype(int)
users_mobile_payment_df['users_id'] = users_mobile_payment_df['users_id'].astype(int)
users_mobile_wallet_df['users_id'] = users_mobile_wallet_df['users_id'].astype(int)
users_mobile_banking_df['users_id'] = users_mobile_banking_df['users_id'].astype(int)
users_trading_df['users_id'] = users_trading_df['users_id'].astype(int)
users_fintech_df['users_id'] = users_fintech_df['users_id'].astype(int)
users_financial_df['users_id'] = users_financial_df['users_id'].astype(int)
users_money_management_df['users_id'] = users_money_management_df['users_id'].astype(int)
users_stocks_df['users_id'] = users_stocks_df['users_id'].astype(int)
users_transaction_df['users_id'] = users_transaction_df['users_id'].astype(int)

### Merge "Tweets" DataFrames With "Deets" DataFrames

- **Merge "tweets" with "deets" using a left join**
- If there is no "deets" information, then "NaN" will show in all "deets" columns

In [299]:
investing_df = pd.merge(
    tweets_investing_df, 
    deets_investing_df, 
    left_on='tweets_id', 
    right_on='deets_tweet_id', 
    how='left')

mobile_payment_df = pd.merge(
    tweets_mobile_payment_df, 
    deets_mobile_payment_df, 
    left_on='tweets_id', 
    right_on='deets_tweet_id', 
    how='left')

mobile_wallet_df = pd.merge(
    tweets_mobile_wallet_df, 
    deets_mobile_wallet_df, 
    left_on='tweets_id', 
    right_on='deets_tweet_id', 
    how='left')

mobile_banking_df = pd.merge(
    tweets_mobile_banking_df, 
    deets_mobile_banking_df, 
    left_on='tweets_id', 
    right_on='deets_tweet_id', 
    how='left')

trading_df = pd.merge(
    tweets_trading_df, 
    deets_trading_df, 
    left_on='tweets_id', 
    right_on='deets_tweet_id', 
    how='left')

fintech_df = pd.merge(
    tweets_fintech_df, 
    deets_fintech_df, 
    left_on='tweets_id', 
    right_on='deets_tweet_id', 
    how='left')

financial_df = pd.merge(
    tweets_financial_df, 
    deets_financial_df, 
    left_on='tweets_id', 
    right_on='deets_tweet_id', 
    how='left')

money_management_df = pd.merge(
    tweets_money_management_df, 
    deets_money_management_df, 
    left_on='tweets_id', 
    right_on='deets_tweet_id', 
    how='left')

stocks_df = pd.merge(
    tweets_stocks_df, 
    deets_stocks_df, 
    left_on='tweets_id', 
    right_on='deets_tweet_id', 
    how='left')

transaction_df = pd.merge(
    tweets_transaction_df, 
    deets_transaction_df, 
    left_on='tweets_id', 
    right_on='deets_tweet_id', 
    how='left')

In [300]:
#make sure length of rows is still correct...should be 5000:
investing_df.shape

(5000, 47)

### Merge "Tweets + Deets" DataFrames With "Users" DataFrames

- **Merge "tweets + deets" DataFrames with "users" using a left join**
- Need to use pd.drop_duplicates() on each "users" DataFrame to ensure that there are no duplicate users that we are merging.

In [326]:
#drop duplicates in each "users" DataFrame and re-assign to a new variable:

users_investing_df_unique = users_investing_df.drop_duplicates(subset=['users_id'])
users_mobile_payment_df_unique = users_mobile_payment_df.drop_duplicates(subset=['users_id'])
users_mobile_wallet_df_unique = users_mobile_wallet_df.drop_duplicates(subset=['users_id'])
users_mobile_banking_df_unique = users_mobile_banking_df.drop_duplicates(subset=['users_id'])
users_trading_df_unique = users_trading_df.drop_duplicates(subset=['users_id'])
users_fintech_df_unique = users_fintech_df.drop_duplicates(subset=['users_id'])
users_financial_df_unique = users_financial_df.drop_duplicates(subset=['users_id'])
users_money_management_df_unique = users_money_management_df.drop_duplicates(subset=['users_id'])
users_stocks_df_unique = users_stocks_df.drop_duplicates(subset=['users_id'])
users_transaction_df_unique = users_transaction_df.drop_duplicates(subset=['users_id'])

In [327]:
investing = pd.merge(
    investing_df, 
    users_investing_df_unique, 
    left_on='tweets_user_id', 
    right_on='users_id', 
    how='left')

mobile_payment = pd.merge(
    mobile_payment_df, 
    users_mobile_payment_df_unique, 
    left_on='tweets_user_id', 
    right_on='users_id', 
    how='left')

mobile_wallet = pd.merge(
    mobile_wallet_df, 
    users_mobile_wallet_df_unique, 
    left_on='tweets_user_id', 
    right_on='users_id', 
    how='left')

mobile_banking = pd.merge(
    mobile_banking_df, 
    users_mobile_banking_df_unique, 
    left_on='tweets_user_id', 
    right_on='users_id', 
    how='left')

trading = pd.merge(
    trading_df, 
    users_trading_df_unique, 
    left_on='tweets_user_id', 
    right_on='users_id', 
    how='left')

fintech = pd.merge(
    fintech_df, 
    users_fintech_df_unique, 
    left_on='tweets_user_id', 
    right_on='users_id', 
    how='left')

financial = pd.merge(
    financial_df, 
    users_financial_df_unique, 
    left_on='tweets_user_id', 
    right_on='users_id', 
    how='left')

money_management = pd.merge(
    money_management_df, 
    users_money_management_df_unique, 
    left_on='tweets_user_id', 
    right_on='users_id', 
    how='left')

stocks = pd.merge(
    stocks_df, 
    users_stocks_df_unique, 
    left_on='tweets_user_id', 
    right_on='users_id', 
    how='left')

transaction = pd.merge(
    transaction_df, 
    users_transaction_df_unique, 
    left_on='tweets_user_id', 
    right_on='users_id', 
    how='left')

- **Check shape to make sure that all tweets are accounted for**
- All "tweets" DataFrames should match the number of rows as the "merged" DataFrames
- There should always be 63 columns in each DataFrame

In [334]:
# Display shape of original "tweets" DataFrames to check merged DataFrames against:

display(
    tweets_investing_df.shape,
    tweets_mobile_payment_df.shape,
    tweets_mobile_wallet_df.shape,
    tweets_mobile_banking_df.shape,
    tweets_trading_df.shape,
    tweets_fintech_df.shape,
    tweets_financial_df.shape,
    tweets_money_management_df.shape,
    tweets_stocks_df.shape,
    tweets_transaction_df.shape
)

(5000, 37)

(1226, 37)

(2035, 37)

(1279, 37)

(5000, 37)

(4965, 37)

(5000, 37)

(4998, 37)

(4876, 37)

(4997, 37)

In [336]:
# Display shape of merged DataFrames 
# All "tweets" DataFrames should match the number of rows as the "merged" DataFrames
# There should always be 63 columns in each DataFrame

display(
    investing.shape,
    mobile_payment.shape,
    mobile_wallet.shape,
    mobile_banking.shape,
    trading.shape,
    fintech.shape,
    financial.shape,
    money_management.shape,
    stocks.shape,
    transaction.shape
)

(5000, 63)

(1226, 63)

(2035, 63)

(1279, 63)

(5000, 63)

(4965, 63)

(5000, 63)

(4998, 63)

(4876, 63)

(4997, 63)

**Yay! Merged DataFrames look good!**

### Concat All "Merged" DataFrames Together Into One Big DataFrame

- Concat all "merged" DataFrames together into one big DataFrame for further cleaning, analysis, and export
- Need to use pd.concat()

In [485]:
df = pd.concat([
    investing,
    mobile_payment,
    mobile_wallet,
    mobile_banking,
    trading,
    fintech,
    financial,
    money_management,
    stocks,
    transaction])

In [486]:
df = df.reset_index(drop = True) 

In [487]:
df.shape

(39376, 63)

In [488]:
#check length against total length of all DataFrames summed:
investing.shape[0] + mobile_payment.shape[0] + mobile_wallet.shape[0] + mobile_banking.shape[0] + trading.shape[0] + fintech.shape[0] + financial.shape[0] + money_management.shape[0] + stocks.shape[0] + transaction.shape[0]

39376

**Yay! Final merged DataFrame matches length!**

___

## Column Cleaning

### Drop columns

**Columns To Drop:**
- **TWEETS**
    - tweets_coordinates - not enough data (86 rows)
    - tweets_geo - not enough data (86 rows)
    - tweets_place - not enough data (445 rows)
    - tweets_contributors - not enough data (0 rows)
    - tweets_user_following - not enough data (0 rows)
    - tweets_user_follow_request_sent - not enough data (0 rows)
    - tweets_user_notifications - not enough data (0 rows)
    - tweets_in_reply_to_status_id - not useful data.
    - tweets_in_reply_to_user_id - not useful data.
    - tweets_in_reply_to_screen_name - not useful data.
    - tweets_source - not data that we really care about.
    - tweets_entities_symbols - out of time to work more with this feature.
    - tweets_entities_user_mentions - out of time to work more with this feature.
    - tweets_entities_urls - out of time to work more with this feature.
    
    
- **DEETS**
    - deets_author_id - this is the same as users_id...no need to have a duplicate column.
    - deets_tweet_id - this is the same as tweets_id...no need to have a duplicate column.
    - deets_lang - we already filtered by "english" when we called tweets. No need for this column.
    - deets_created_at - we already have tweets_created_at - this is the same thing. No need for this column.


- **USERS**
    - users_url_start - after analyzing, we aren't sure what this data really means, so we don't want to use it.
    - users_url_end - after analyzing, we aren't sure what this data really means, so we don't want to use it.

In [492]:
df = df.drop([
    'tweets_coordinates',
    'tweets_geo',
    'tweets_place',
    'tweets_contributors',
    'tweets_user_following',
    'tweets_user_follow_request_sent',
    'tweets_user_notifications',
    'tweets_in_reply_to_status_id',
    'tweets_in_reply_to_user_id',
    'tweets_in_reply_to_screen_name',
    'tweets_source',
    'tweets_entities_symbols',
    'tweets_entities_user_mentions',
    'tweets_entities_urls',
    'deets_tweet_id',
    'deets_author_id',
    'deets_lang',
    'deets_created_at',
    'users_url_start',
    'users_url_end'], 
    axis=1)

**Columns To Convert:**
- **TWEETS**
    - tweets_created_at - str to datetime
    - tweets_entities_symbols - just get 'text' out of this dictionary (IF TIME PERMITS...OTHERWISE DELETE. **DELETED.**)
    - tweets_entities_user_mentions - lots of data in here sometimes...convert this into a bool. (IF TIME PERMITS...OTHERWISE DELETE. **DELETED.**)
    - tweets_entities_urls - lots of data in here sometimes...convert this into a bool. (IF TIME PERMITS...OTHERWISE DELETE. **DELETED.**)


- **DEETS**
    - deets_possibly_sensitive - str to bool


- **USERS**
    - users_created_at - str to datetime

In [493]:
#tweets
df['tweets_created_at'] = pd.to_datetime(df['tweets_created_at'])

#deets
df['deets_possibly_sensitive'] = df['deets_possibly_sensitive'].astype(bool)

#users
df['users_created_at'] = pd.to_datetime(df['users_created_at'])

**Feature Engineering**
- **TWEETS**
    - NONE
    
    
- **DEETS**
    - deets_text_word_length - create column of number of words in tweet. 
    - deets_text_character_length - create column of number of characters in tweet.

    
- **USERS**
    - users_description_word_length - create column of number of words in user description.
    - users_description_character_length - create column of number of characters in user description.

In [494]:
df['deets_text_word_length'] = df['deets_text'].str.split().str.len()
df['deets_text_character_length'] = df['deets_text'].str.len()
df['users_description_word_length'] = df['users_description'].str.split().str.len()
df['users_description_character_length'] = df['users_description'].str.len()

In [495]:
df.head()

tweets_created_at            tweets_id  tweets_is_quote_status  tweets_favorited  tweets_favorite_count  tweets_retweeted  tweets_retweet_count      tweets_user_id      tweets_user_name tweet_user_screen_name tweets_user_location                            tweets_user_description  tweets_user_friends_count  tweets_user_followers_count  tweets_user_favourites_count  tweets_user_geo_enabled  tweets_user_verified  tweets_user_statuses_count tweets_user_profile_background_color tweets_user_profile_text_color  tweets_user_profile_use_background_image  tweets_user_has_extended_profile  tweets_user_default_profile_image  deets_possibly_sensitive                                       deets_source                                         deets_text  deets_reply_count  deets_like_count  deets_quote_count      users_id          users_created_at                                  users_description            users_name   users_username users_verified users_protected  \
0 2020-01-31 21:03:41+00:00  1223351165523578885                   False             False                  12761             False                  3014           284278132         Morgan Housel           morganhousel                                                             @collabfund                        487                       129374                         54543                    False                  True                       14545                               131516                         333333                                      True                              True                              False                     False  <a href="http://twitter.com/download/iphone" r...                Investing.  https://t.co/qQQtVyOLid              195.0           14083.0              444.0  2.842781e+08 2011-04-19 00:24:40+00:00                                        @collabfund         Morgan Housel     morganhousel           True           False   
1 2020-01-31 03:05:28+00:00  1223079825566359552                   False             False                  21733             False                 12434            35218566         Ari Fleischer           AriFleischer                       Wikipedia has my place of birth wrong. Former ...                       1462                       369929                           296                     True                  True                       16723                               C0DEED                         333333                                      True                             False                               True                     False  <a href="http://twitter.com/download/iphone" r...  Did you know that the Obama-Biden administrati...              621.0           22082.0              715.0  3.521857e+07 2009-04-25 14:11:19+00:00  Wikipedia has my place of birth wrong. Former ...         Ari Fleischer     AriFleischer           True           False   
2 2020-01-31 17:02:41+00:00  1223290518689452032                   False             False                   2337             False                   654            11134252                   GOP                    GOP                       Sign up to get official updates from the Repub...                       1668                      2229921                           278                    False                  True                       37946                               F0F0F0                         000000                                     False                             False                              False                     False  <a href="https://sproutsocial.com" rel="nofoll...  “Americans are investing in a party and a pres...              439.0            2355.0               33.0  1.113425e+07 2007-12-13 17:15:41+00:00  Sign up to get official updates from the Repub...                   GOP              GOP           True           False   
3 2020-02-01 21:03:46+00:00  1223713574083756032                   False             False    

In [496]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39376 entries, 0 to 39375
Data columns (total 47 columns):
tweets_created_at                           39376 non-null datetime64[ns, UTC]
tweets_id                                   39376 non-null int64
tweets_is_quote_status                      39376 non-null bool
tweets_favorited                            39376 non-null bool
tweets_favorite_count                       39376 non-null int64
tweets_retweeted                            39376 non-null bool
tweets_retweet_count                        39376 non-null int64
tweets_user_id                              39376 non-null int64
tweets_user_name                            39376 non-null object
tweet_user_screen_name                      39376 non-null object
tweets_user_location                        39376 non-null object
tweets_user_description                     39376 non-null object
tweets_user_friends_count                   39376 non-null int64
tweets_user_followers_count    

___

## Filter DataFrame Rows By "Startup" Criteria

- Our business case is for a financial tech startup company.
- This startup company is likely to only have a few thousand followers at best.
- Tweets by Twitter influencers with millions of followers will be weighted too heavily in predicting more retweets...
- Based on our research, we have found that startup companies tend to have a follower count between 1,000 and 10,000.
- **So, to control for this, we are only interested in tweets by users that have a follower count between 1,000 and 10,000, like our startup company.**

In [497]:
#filter tweets by users with follower count between 1,000 and 10,000
df = df.loc[(df['users_followers_count'] >= 1000) & (df['users_followers_count'] <= 10000)]

In [498]:
#how many rows of data did we lose?
df.shape

(12070, 47)

- After filtering, we lost about 2/3 of our data. (39376 rows --> 12070 rows)
- But we still have over 10,000 rows to work with, this should still be adequate.

## Imputing Values As Necessary

### Figure out where NaN values are in our data

**There are 331 NaN values in:**
- deets_source
- deets_text
- deets_reply_count
- deets_like_count
- deets_quote_count
- deets_text_word_length
- deets_text_character_length


**Losing 331 rows isn't too much of a loss. We will eliminate these ^ rows of NaN data.**

In [499]:
df.isna().sum()

tweets_created_at                             0
tweets_id                                     0
tweets_is_quote_status                        0
tweets_favorited                              0
tweets_favorite_count                         0
tweets_retweeted                              0
tweets_retweet_count                          0
tweets_user_id                                0
tweets_user_name                              0
tweet_user_screen_name                        0
tweets_user_location                          0
tweets_user_description                       0
tweets_user_friends_count                     0
tweets_user_followers_count                   0
tweets_user_favourites_count                  0
tweets_user_geo_enabled                       0
tweets_user_verified                          0
tweets_user_statuses_count                    0
tweets_user_profile_background_color          0
tweets_user_profile_text_color                0
tweets_user_profile_use_background_image

In [502]:
df = df.dropna(how='any', axis=0) 

In [503]:
df.isna().sum()

tweets_created_at                           0
tweets_id                                   0
tweets_is_quote_status                      0
tweets_favorited                            0
tweets_favorite_count                       0
tweets_retweeted                            0
tweets_retweet_count                        0
tweets_user_id                              0
tweets_user_name                            0
tweet_user_screen_name                      0
tweets_user_location                        0
tweets_user_description                     0
tweets_user_friends_count                   0
tweets_user_followers_count                 0
tweets_user_favourites_count                0
tweets_user_geo_enabled                     0
tweets_user_verified                        0
tweets_user_statuses_count                  0
tweets_user_profile_background_color        0
tweets_user_profile_text_color              0
tweets_user_profile_use_background_image    0
tweets_user_has_extended_profile  

In [504]:
df.shape

(11739, 47)

## Export Final DataFrame To CSV

In [505]:
df.to_csv('Twitter_Data.csv', encoding='utf-8')